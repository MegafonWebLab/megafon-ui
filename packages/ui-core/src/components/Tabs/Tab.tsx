import * as React from 'react';
import PropTypes from 'prop-types';

export interface ITabProps {
    /** Заголовок таба */
    title?: string;
    /** Иконка таба */
    icon?: React.ReactNode;
    /** Ссылка */
    href?: string;
    /** Функция рендера компонента-обертки для заголовка и иконки */
    renderTabWrapper?: (tab: React.ReactNode) => React.ReactNode;
}

const Tab: React.FC<ITabProps> = ({ children }) => <>{children}</>;

Tab.propTypes = {
    title: PropTypes.string,
    icon: PropTypes.node,
    href: PropTypes.string,
    renderTabWrapper: PropTypes.func,
};

export default Tab;
