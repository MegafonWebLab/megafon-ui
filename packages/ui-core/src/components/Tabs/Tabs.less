@import '~styles/base.less';

@b: e('.mfui-beta-tabs');
@swiperWrapper: e('.swiper-wrapper');

@{b} {
    &__swiper {
        .fontFamily();
        .h5();
        position: relative;
        overflow: hidden;
        display: flex;

        &:before,
        &:after {
            content: '';
            position: absolute;
            z-index: 1;
            top: 0;
            bottom: 0;
            width: 64px;
        }

        &:before {
            left: 0;
            background: linear-gradient(to right, @clearWhite, rgba(@clearWhite, 0));
        }

        &:after {
            right: 0;
            background: linear-gradient(to left, @clearWhite, rgba(@clearWhite, 0));
        }
    }

    &__swiper_beginning {
        &:before {
            display: none;
        }
    }

    &__swiper_end {
        &:after {
            display: none;
        }
    }

    @{swiperWrapper} {
        position: relative;
        z-index: 0;
        display: flex;
        max-width: 100%;

        @media @mobileSM {
            width: 100%;
        }
    }

    &__slide {
        -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
        flex-shrink: 0;

        @media @mobileSM {
            flex-grow: 1;
        }
    }

    &__tab-inner {
        color: @freshAsphalt;
        fill: @freshAsphalt;
        white-space: nowrap;
        box-sizing: content-box;
        display: flex;
        height: 32px;
        align-items: center;
        justify-content: center;
        cursor: pointer;
        text-decoration: none;

        &:hover {
            color: @green;
            fill: @green;
        }
    }

    &__tab-inner_current {
        color: @green;
        fill: @green;
    }

    &__tab-icon {
        width: 32px;
        height: 32px;

        &:not(:last-child) {
            margin-right: 8px;
        }

        svg {
            width: 100%;
            height: 100%;
        }
    }

    &__panel {
        display: none;
    }

    &__panel_current {
        display: block;
    }

    &__underline {
        position: absolute;
        bottom: 0;
        left: 0;
        height: 2px;
        background-color: @green;
        transition-duration: 0.3s;
    }

    &__arrow {
        position: absolute;
        z-index: 2;
        top: 50%;
        transform: translateY(-50%);
        display: block;
        width: 16px;
        height: 16px;
        border-radius: 50%;
        background-color: @freshAsphalt;
        fill: @clearWhite;
        cursor: pointer;

        @media @mobile {
            display: none;
        }
    }

    &__arrow_prev {
        left: 0;
    }

    &__arrow_next {
        right: 0;
    }

    &__arrow_hide {
        display: none;
    }

    // mod size small

    &_size_small {
        @{b}__slide:not(:last-child) @{b}__tab {
            padding-right: 40px;

            @media @mobileSM {
                padding-right: 32px;
            }
        }

        @{b}__tab-inner {
            height: 24px;
            padding: 7px 0;
            justify-content: start;

            @media @mobileSM {
                padding: 3px 0;
            }
        }

        @{b}__tab-icon {
            width: 24px;
            height: 24px;
        }
    }

    // mod size medium

    &_size_medium {
        @{b}__tab-inner {
            padding: 7px 24px;

            @media @mobileSM {
                padding: 3px 16px;
            }
        }
    }

    // mod size large

    &_size_large {
        @{b}__swiper {
            .h3();
        }

        @{b}__tab-inner {
            padding: 11px 24px;

            @media @mobileSM {
                padding: 7px 16px;
            }
        }
    }

    // mod hAlign center

    &_h-align_center {
        @{b}__swiper {
            justify-content: center;
        }

        @{b}__tab {
            border-bottom: 1px solid @spbSky2;
        }
    }

    &_h-align_left {
        @{b}__swiper-wrapper {
            position: relative;

            &:before {
                content: '';
                position: absolute;
                left: 0;
                bottom: 0;
                width: 100%;
                height: 1px;
                background-color: @spbSky2;
            }
        }
    }

    // mod indents

    &_indents {
        @{b}__swiper {
            @media @mobileB {
                margin: 0 -@contentHorizontalIndentMobileB;
                padding: 0 @contentHorizontalIndentMobileB;
            }

            @media @mobileSM {
                margin: 0 -@contentHorizontalIndentMobileSM;
                padding: 0 @contentHorizontalIndentMobileSM;
            }
        }
    }

    // mod sticky

    &_sticky {
        @{b}__swiper-wrapper {
            position: fixed;
            top: 0;
            right: 0;
            left: 0;
            z-index: 100;
            background-color: @clearWhite;
        }
    }

    // mod color green

    &_tab-color_green:not(&_sticky) {
        @{b}__swiper {
            &:before {
                left: 0;
                background: linear-gradient(to right, @green, transparent);
            }

            &:after {
                right: 0;
                background: linear-gradient(to left, @green, transparent);
            }
        }

        @{b}__tab {
            border-color: fade(@spbSky1, 50%);
        }

        @{b}__tab-inner {
            color: fade(@clearWhite, 50%);
            fill: fade(@clearWhite, 50%);

            &:hover {
                color: @clearWhite;
                fill: @clearWhite;
            }
        }

        @{b}__tab-inner_current {
            color: @clearWhite;
            fill: @clearWhite;
        }

        @{b}__underline {
            background-color: @clearWhite;
        }

        @{b}__arrow {
            background-color: @clearWhite;
            fill: @green;
        }
    }
}
