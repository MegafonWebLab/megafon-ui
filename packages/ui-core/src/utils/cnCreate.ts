import classnames from 'classnames';

interface IModificators {
    [key: string]: boolean | string | undefined | null;
}

interface IClassSet {
    [key: string]: boolean;
}

type CustomClassNameType = string;
type CustomClassNamesArrayType = string[];
type CustomClassNamesType = CustomClassNameType | CustomClassNamesArrayType | undefined;

type ElementNameType = string | IModificators | CustomClassNamesArrayType | CustomClassNamesType;

const getCustomClassName = (first?: ElementNameType, second?: ElementNameType, third?: CustomClassNamesType) => {
    if (Array.isArray(first)) {
        return first.filter(f => typeof f === 'string').join(' ');
    }
    if (Array.isArray(second)) {
        return second.filter(f => typeof f === 'string').join(' ');
    }
    if (typeof second === 'string') {
        return second;
    }
    if (Array.isArray(third)) {
        return third.filter(f => typeof f === 'string').join(' ');
    }
    if (typeof third === 'string') {
        return third;
    }

    return '';
};

/**
 * Бэм генератор
 * @param {String} blockName - название блока
 * @returns {function} - возвращает функцию с тремя возожными параметрами
 * @param {String} elementName - название элемента
 * @param {Object} modificatorsObject - объект с модификаторами
 * @param {String} customClassNames - названия кастомных классов
 * @returns {String}
 */
function cnCreate(blockName: string) {
    function getElement(): string;
    function getElement(elementName: ElementNameType): string;
    function getElement(modificatorsObject: ElementNameType): string;
    function getElement(customClassNames: CustomClassNamesArrayType): string;
    function getElement(elementName: ElementNameType, modificatorsObject: ElementNameType): string;
    function getElement(elementName: ElementNameType, customClassNames: CustomClassNamesType): string;
    function getElement(modificatorsObject: ElementNameType, customClassNames: CustomClassNamesType): string;
    function getElement(
        elementName: ElementNameType,
        modificatorsObject: ElementNameType,
        customClassNames: CustomClassNamesType
    ): string;
    function getElement(
        elementName?: ElementNameType,
        modificatorsObject?: ElementNameType,
        customClassNames?: CustomClassNamesType
    ): string {
        const prefix = typeof elementName === 'string' ? elementName : '';
        const className = getCustomClassName(elementName, modificatorsObject, customClassNames);
        const secondMods = !Array.isArray(modificatorsObject) && typeof modificatorsObject !== 'string'
            ? modificatorsObject
            : undefined;
        const params = typeof elementName !== 'string' && !Array.isArray(elementName)
            ? elementName
            : secondMods;

        if (!params || typeof params === 'object' && Object.keys(params).length === 0) {
            return classnames(`${blockName}${prefix ? '__' + prefix : ''}`, className);
        }

        const classParams: IClassSet = {};
        let prefixKey = '', withoutPrefix = '';

        for (const key in params) {
            if (typeof params[key] === 'boolean' && params[key]) {
                prefixKey = `__${prefix}_${key}`;
                classParams[
                    `${blockName}${prefix ? prefixKey : '_' + key}`
                ] = true;
            } else if (typeof params[key] === 'string') {
                prefixKey = `__${prefix}_${key}_${params[key]}`;
                withoutPrefix = `_${key}_${params[key]}`;
                classParams[
                    `${blockName}${prefix ? prefixKey : withoutPrefix}`
                ] = true;
            }
        }

        return prefix === ''
            ? classnames(blockName, classParams, className)
            : classnames(`${blockName}__${prefix}`, classParams, className);
    }

    return getElement;
}

export default cnCreate;
