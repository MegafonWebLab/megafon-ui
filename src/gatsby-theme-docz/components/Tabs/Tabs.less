.docz-tabs {
    &__tab-list {
        position: relative;
        display: flex;
        padding: 0;
        border-bottom: 1px solid #EDEDED;
    }

    &__tab {
        display: block;
        background: none;
        border: none;
        padding: 16px;

        font-size: 15px;
        line-height: 24px;
        font-weight: 500;
        outline: none;
        cursor: pointer;

        &[aria-selected="true"] {
            color: #00B956;
        }
    }

    &__underline {
        position: absolute;
        bottom: -1px;
        left: 0;
        width: 100%;
        height: 2px;
    }

    &__underline-swipe {
        width: 0;
        height: inherit;
        background: #00B956;
        transition: 0.3s;
    }
}
