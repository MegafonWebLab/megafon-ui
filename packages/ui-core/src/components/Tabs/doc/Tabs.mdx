---
name: Tabs
menu: Компоненты
route: components/tabs
---

import { Props } from 'docz';
import Tabs from '../Tabs';
import Tab from '../Tab';
import Example from './Tabs.example.mdx';

# Tabs

<CopyToClipboardBox text="import { Tabs, Tab } from '@megafon/ui-core';" />

<DoczTabs>
    <Example />
    <div>
        <Props of={Tabs} />
        <h2>Таб</h2>
        <Props of={Tab} />
    </div>
</DoczTabs>
