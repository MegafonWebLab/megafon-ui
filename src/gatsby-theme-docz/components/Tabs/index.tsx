import React from 'react';
import { Tabs, TabList, Tab, TabPanel } from "react-tabs";
import { cnCreate } from '@megafon/ui-core';
import './Tabs.less';

const TextMap = ['Примеры и код', 'Свойства и методы', 'Правила использования'];
const cn = cnCreate('docz-tabs');

export const DoczTabs: React.FC = ({ children }) => {
    const isList = Array.isArray(children);
    const tabs = React.useMemo(() => Array.isArray(children) ? [...children].slice(0, TextMap.length) : [], [children]);
    const tabsRef = React.useRef<number[]>([]);
    const [state, setState] = React.useState({ underlineWidth: 0, underlineTransform: 0 });

    const setTabRef = React.useCallback((n) => {
        if (n) {
            tabsRef.current.push(n.node.clientWidth);
        }
    }, []);

    const handleSelect = React.useCallback((index: number) => {
        const underlineTransform = [...tabsRef.current].splice(0, index).reduce((a,b) => a + b, 0);
        setState({ underlineWidth: tabsRef.current[index], underlineTransform });
    }, []);

    React.useEffect(() => {
        if (!isList) {
            console.error('DoczTabs: children must be array of components');
        }
    }, [isList]);

    React.useEffect(() => {
        if (tabsRef.current[0]) {
            setState({ underlineWidth: tabsRef.current[0], underlineTransform: 0 });
        }
    }, []);

    return (
        <Tabs className={cn()} onSelect={handleSelect}>
            <TabList className={cn('tab-list')}>
                {tabs.map((_, i) =>
                    <Tab
                        key={i}
                        ref={setTabRef}
                        className={cn('tab')}
                    >
                        {TextMap[i]}
                    </Tab>
                )}
                <div className={cn('underline')}>
                    <div
                        className={cn('underline-swipe')}
                        style={{
                            width: `${state.underlineWidth}px`,
                            transform: `translateX(${state.underlineTransform}px)`,
                        }}
                    />
                </div>
            </TabList>
            {tabs.map((tab, i) => <TabPanel key={i}>{tab}</TabPanel>)}
        </Tabs>
    );
}
