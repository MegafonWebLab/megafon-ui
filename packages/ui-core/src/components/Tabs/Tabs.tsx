import * as React from 'react';
import PropTypes from 'prop-types';
import SwiperCore from 'swiper';
import { Swiper, SwiperSlide } from 'swiper/react';
import throttle from 'lodash.throttle';
import cnCreate from 'utils/cnCreate';
import './Tabs.less';
import { ITabProps } from './Tab';
import ArrowLeft from 'icons/System/16/Arrow_left_16.svg';
import ArrowRight from 'icons/System/16/Arrow_right_16.svg';

export const TabSize = {
    SMALL: 'small',
    MEDIUM: 'medium',
    LARGE:  'large',
} as const;

export const TabHAlign = {
    LEFT: 'left',
    CENTER: 'center',
} as const;

export const TabColorTheme = {
    WHITE: 'white',
    GREEN: 'green',
} as const;

type TabSizeType = typeof TabSize[keyof typeof TabSize];
type TabHAlignType = typeof TabHAlign[keyof typeof TabHAlign];
type TabColorThemeType = typeof TabColorTheme[keyof typeof TabColorTheme];

export interface ITabsProps {
    /** Дополнительный класс для корневого элемента */
    className?: string;
    /** Дополнительные классы для корневого и внутренних элементов */
    classes?: {
        root?: string;
        innerIndents?: string;
    };
    /** Размер табов */
    size?: TabSizeType;
    /** Горизонтальное выравнивание */
    hAlign?: TabHAlignType;
    /** Цветовая тема табов для определенных фонов */
    tabColorTheme?: TabColorThemeType;
    /** Фиксация табов у верхней границы окна */
    sticky?: boolean;
    /** Индекс активного таба (включает режим управления табами снаружи) */
    currentIndex?: number;
    /** Индекс активного таба по умолчанию (не работает в режиме управления табами снаружи) */
    defaultIndex?: number;
    /** Обработчика клика по табам */
    onTabClick?: (index: number) => void;
    children: Array<React.ReactElement<ITabProps>>;
}

const cn = cnCreate('mfui-beta-tabs');
const Tabs: React.FC<ITabsProps> = ({
    className,
    classes: { root: rootClass, innerIndents: innerIndentsClass } = {},
    size = 'medium',
    hAlign = 'left',
    tabColorTheme = 'white',
    sticky = false,
    defaultIndex = 0,
    currentIndex: outerIndex,
    children,
    onTabClick,
}) => {
    const tabsRef = React.useRef<HTMLDivElement[]>([]);
    const rootRef = React.useRef<HTMLDivElement>(null);
    const tabListRef = React.useRef<HTMLDivElement>(null);

    const [swiperInstance, setSwiperInstance] = React.useState<SwiperCore>();
    const [isBeginning, setBeginning] = React.useState(true);
    const [isEnd, setEnd] = React.useState(false);

    const [innerIndex, setInnerIndex] = React.useState(defaultIndex);
    const currentIndex = outerIndex === undefined ? innerIndex : outerIndex;

    const [underlineWidth, setUnderlineWidth] = React.useState(0);
    const [underlineTranslate, setUnderlineTranslate] = React.useState(0);
    const [underlineTransition, setUnderlineTransition] = React.useState(
        'none'
    );

    const [tabListHeight, setTabListHeight] = React.useState<number | 'auto'>(
        'auto'
    );

    const [isSticky, setSticky] = React.useState(false);
    const [stickyOffset, setStickyOffset] = React.useState({
        left: 0,
        right: 0,
    });

    const setTabRef = React.useCallback((tab: HTMLDivElement) => {
        tab && tabsRef.current.push(tab);
    }, []);

    const calculateUnderline = React.useCallback(() => {
        if (!tabsRef.current.length) {
            return;
        }

        const tabNodeChild = tabsRef.current[currentIndex].firstElementChild;
        const { clientWidth = 0 } = (tabNodeChild as HTMLDivElement) || {};
        const translate = [...tabsRef.current]
            .splice(0, currentIndex)
            .reduce((accWidth, node) => {
                const { width } = node.getBoundingClientRect();

                return accWidth + width;
            }, 0);

        setUnderlineWidth(clientWidth);
        setUnderlineTranslate(translate);
    }, [currentIndex]);

    const calculateSticky = React.useCallback(() => {
        if (!sticky || !rootRef.current || !tabListRef.current) {
            return;
        }

        const { top, left, right } = rootRef.current.getBoundingClientRect();
        const documentWidth = document.documentElement.clientWidth;
        const listHeight = tabListRef.current.clientHeight;
        const rootHeight = rootRef.current.clientHeight;
        const contentHeight = rootHeight - listHeight;
        const needSetSticky = top < 0 && Math.abs(top) < contentHeight;
        const stickyLeftOffset = needSetSticky ? left : 0;
        const stickyRightOffset = needSetSticky ? documentWidth - right : 0;

        setTabListHeight(needSetSticky ? listHeight : 'auto');
        setSticky(needSetSticky);
        setStickyOffset({ left: stickyLeftOffset, right: stickyRightOffset });
    }, [sticky]);

    const handleTabInnerClick = React.useCallback((index: number) => () => {
        setUnderlineTransition('all');

        onTabClick && onTabClick(index);
        if (outerIndex === undefined) {
            setInnerIndex(index);
        }
    }, [onTabClick, outerIndex]);

    const handleSwiper = React.useCallback((swiper: SwiperCore) => {
        setSwiperInstance(swiper);
    }, []);

    const handlePrevArrowClick = React.useCallback(() => {
        swiperInstance?.slidePrev();
    }, [swiperInstance]);

    const handleNextArrowClick = React.useCallback(() => {
        swiperInstance?.slideNext();
    }, [swiperInstance]);

    const renderTab = React.useCallback((
        index: number,
        title?: string,
        icon?: React.ReactNode,
        href?: string
    ) => {
        const ElementType = href ? 'a' : 'div';

        return (
            <ElementType
                href={href}
                className={cn('tab-inner', {
                    current: currentIndex === index,
                })}
                onClick={handleTabInnerClick(index)}
            >
                {!!icon && <div className={cn('tab-icon')}>{icon}</div>}
                {!!title && <div className={cn('tab-title')}>{title}</div>}
            </ElementType>
        );
    }, [handleTabInnerClick, currentIndex]);

    const renderTabs = React.useCallback(() =>
        React.Children.map(children, (child, i) => {
            const {
                props: { title, icon, href, renderTabWrapper },
            } = child;
            const tab = renderTab(i, title, icon, href);

            return (
                <SwiperSlide className={cn('slide')}>
                    <div className={cn('tab')} ref={setTabRef}>
                        {renderTabWrapper ? renderTabWrapper(tab) : tab}
                    </div>
                </SwiperSlide>
            );
    }), [renderTab, children]);

    const renderPanels = React.useCallback(() =>
        React.Children.map(children, (child, i) => {
            return (
                <div
                    className={cn('panel', {
                        current: currentIndex === i,
                    })}
                >
                    {child}
                </div>
            );
    }), [children, currentIndex]);

    const handleReachBeginning = React.useCallback((swiper: SwiperCore) => {
        setBeginning(swiper.isBeginning);
    }, []);

    const handleReachEnd = React.useCallback((swiper: SwiperCore) => {
        setEnd(swiper.isEnd);
    }, []);

    const handleFromEdge = React.useCallback((swiper: SwiperCore) => {
        setBeginning(swiper.isBeginning);
        setEnd(swiper.isEnd);
    }, []);

    React.useEffect(() => {
        const handleScroll = throttle(calculateSticky, 50);

        document.addEventListener('scroll', handleScroll);

        return () => {
            document.removeEventListener('scroll', handleScroll);
        };
    }, [calculateSticky]);

    React.useEffect(() => {
        const handleResize = throttle(() => {
            calculateUnderline();
            calculateSticky();
        }, 300);

        calculateUnderline();

        window.addEventListener('resize', handleResize);

        return () => {
            window.removeEventListener('resize', handleResize);
        };
    }, [calculateUnderline, calculateSticky]);

    React.useEffect(() => {
        if (!swiperInstance) {
            return;
        }

        setBeginning(swiperInstance.isBeginning);
        setEnd(swiperInstance.isEnd);
    }, [swiperInstance]);

    return (
        <div
            className={cn(
                {
                    size,
                    'h-align': hAlign,
                    'tab-color': tabColorTheme,
                    indents: !innerIndentsClass,
                    sticky: isSticky,
                },
                [className, rootClass]
            )}
            ref={rootRef}
        >
            <div
                ref={tabListRef}
                style={{ height: tabListHeight }}
            >
                <div
                    className={cn('swiper-wrapper')}
                    style={{
                        paddingLeft: stickyOffset.left,
                        paddingRight: stickyOffset.right,
                    }}
                >
                    <Swiper
                        simulateTouch={false}
                        className={cn(
                            'swiper',
                            {
                                beginning: isBeginning,
                                end: isEnd,
                            },
                            [innerIndentsClass]
                        )}
                        slidesPerView="auto"
                        initialSlide={currentIndex}
                        onSwiper={handleSwiper}
                        onReachBeginning={handleReachBeginning}
                        onReachEnd={handleReachEnd}
                        onFromEdge={handleFromEdge}
                    >
                        {renderTabs()}
                        <div
                            className={cn('underline')}
                            slot="wrapper-start"
                            style={{
                                width: `${underlineWidth}px`,
                                transform: `translateX(${underlineTranslate}px)`,
                                transitionProperty: underlineTransition,
                            }}
                        />
                        <ArrowLeft
                            className={cn('arrow', {
                                prev: true,
                                hide: isBeginning,
                            })}
                            onClick={handlePrevArrowClick}
                        />
                        <ArrowRight
                            className={cn('arrow', { next: true, hide: isEnd })}
                            onClick={handleNextArrowClick}
                        />
                    </Swiper>
                </div>
            </div>
            {renderPanels()}
        </div>
    );
};

Tabs.propTypes = {
    className: PropTypes.string,
    classes: PropTypes.shape({
        root: PropTypes.string,
        innerIndents: PropTypes.string,
    }),
    size: PropTypes.oneOf(Object.values(TabSize)),
    hAlign: PropTypes.oneOf(Object.values(TabHAlign)),
    tabColorTheme: PropTypes.oneOf(Object.values(TabColorTheme)),
    sticky: PropTypes.bool,
    currentIndex: PropTypes.number,
    defaultIndex: PropTypes.number,
    onTabClick: PropTypes.func,
};

export default Tabs;
