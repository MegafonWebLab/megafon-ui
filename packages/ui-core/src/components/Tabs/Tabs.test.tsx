import * as React from 'react';
import { shallow } from 'enzyme';
import cnCreate from 'utils/cnCreate';
import Tabs, { TabSize, TabHAlign, TabColorTheme } from './Tabs';
import Tab from './Tab';
import Balance from 'icons/Basic/24/Balance_24.svg';

const renderTabWrapper = (tab: React.ReactNode) => {
    return <div className="tab-wrapper">{tab}</div>;
};

const cn = cnCreate('.mfui-beta-tabs');
describe('<Tabs />', () => {
    afterAll(() => jest.clearAllMocks());

    it('should render with default props', () => {
        const wrapper = shallow(
            <Tabs>
                <Tab>1</Tab>
                <Tab>2</Tab>
            </Tabs>
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should render with props', () => {
        const wrapper = shallow(
            <Tabs
                className="className"
                classes={{
                    root: 'rootClass',
                    innerIndents: 'innerIndentsClass',
                }}
                size={TabSize.SMALL}
                hAlign={TabHAlign.CENTER}
                tabColorTheme={TabColorTheme.GREEN}
                sticky
                defaultIndex={1}
            >
                <Tab title="title 1" icon={<Balance />} href="www.test.com">
                    1
                </Tab>
                <Tab
                    title="title 2"
                    renderTabWrapper={renderTabWrapper}
                >
                    2
                </Tab>
            </Tabs>
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('chould call onTabClick', () => {
        const mockHandleTabClick = jest.fn();
        const wrapper = shallow(
            <Tabs onTabClick={mockHandleTabClick}>
                <Tab>1</Tab>
                <Tab>2</Tab>
            </Tabs>
        );

        wrapper.find(cn('tab-inner')).last().simulate('click');

        expect(mockHandleTabClick).toBeCalledWith(1);
    });
});
