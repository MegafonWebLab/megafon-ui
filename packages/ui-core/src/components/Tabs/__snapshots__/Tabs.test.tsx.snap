// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> should render with default props 1`] = `
<div
  className="mfui-beta-tabs mfui-beta-tabs_size_medium mfui-beta-tabs_h-align_left mfui-beta-tabs_tab-color_white mfui-beta-tabs_indents"
>
  <div
    style={
      Object {
        "height": "auto",
      }
    }
  >
    <div
      className="mfui-beta-tabs__swiper-wrapper"
      style={
        Object {
          "paddingLeft": 0,
          "paddingRight": 0,
        }
      }
    >
      <Swiper
        className="mfui-beta-tabs__swiper mfui-beta-tabs__swiper_beginning"
        initialSlide={0}
        onFromEdge={[Function]}
        onReachBeginning={[Function]}
        onReachEnd={[Function]}
        onSwiper={[Function]}
        simulateTouch={false}
        slidesPerView="auto"
      >
        <SwiperSlide
          className="mfui-beta-tabs__slide"
          key=".0"
        >
          <div
            className="mfui-beta-tabs__tab"
          >
            <div
              className="mfui-beta-tabs__tab-inner mfui-beta-tabs__tab-inner_current"
              onClick={[Function]}
            />
          </div>
        </SwiperSlide>
        <SwiperSlide
          className="mfui-beta-tabs__slide"
          key=".1"
        >
          <div
            className="mfui-beta-tabs__tab"
          >
            <div
              className="mfui-beta-tabs__tab-inner"
              onClick={[Function]}
            />
          </div>
        </SwiperSlide>
        <div
          className="mfui-beta-tabs__underline"
          slot="wrapper-start"
          style={
            Object {
              "transform": "translateX(0px)",
              "transitionProperty": "none",
              "width": "0px",
            }
          }
        />
        <test-file-stub
          className="mfui-beta-tabs__arrow mfui-beta-tabs__arrow_prev mfui-beta-tabs__arrow_hide"
          onClick={[Function]}
        />
        <test-file-stub
          className="mfui-beta-tabs__arrow mfui-beta-tabs__arrow_next"
          onClick={[Function]}
        />
      </Swiper>
    </div>
  </div>
  <div
    className="mfui-beta-tabs__panel mfui-beta-tabs__panel_current"
    key=".0"
  >
    <Tab>
      1
    </Tab>
  </div>
  <div
    className="mfui-beta-tabs__panel"
    key=".1"
  >
    <Tab>
      2
    </Tab>
  </div>
</div>
`;

exports[`<Tabs /> should render with props 1`] = `
<div
  className="mfui-beta-tabs mfui-beta-tabs_size_small mfui-beta-tabs_h-align_center mfui-beta-tabs_tab-color_green className rootClass"
>
  <div
    style={
      Object {
        "height": "auto",
      }
    }
  >
    <div
      className="mfui-beta-tabs__swiper-wrapper"
      style={
        Object {
          "paddingLeft": 0,
          "paddingRight": 0,
        }
      }
    >
      <Swiper
        className="mfui-beta-tabs__swiper mfui-beta-tabs__swiper_beginning innerIndentsClass"
        initialSlide={1}
        onFromEdge={[Function]}
        onReachBeginning={[Function]}
        onReachEnd={[Function]}
        onSwiper={[Function]}
        simulateTouch={false}
        slidesPerView="auto"
      >
        <SwiperSlide
          className="mfui-beta-tabs__slide"
          key=".0"
        >
          <div
            className="mfui-beta-tabs__tab"
          >
            <a
              className="mfui-beta-tabs__tab-inner"
              href="www.test.com"
              onClick={[Function]}
            >
              <div
                className="mfui-beta-tabs__tab-icon"
              >
                <test-file-stub />
              </div>
              <div
                className="mfui-beta-tabs__tab-title"
              >
                title 1
              </div>
            </a>
          </div>
        </SwiperSlide>
        <SwiperSlide
          className="mfui-beta-tabs__slide"
          key=".1"
        >
          <div
            className="mfui-beta-tabs__tab"
          >
            <div
              className="tab-wrapper"
            >
              <div
                className="mfui-beta-tabs__tab-inner mfui-beta-tabs__tab-inner_current"
                onClick={[Function]}
              >
                <div
                  className="mfui-beta-tabs__tab-title"
                >
                  title 2
                </div>
              </div>
            </div>
          </div>
        </SwiperSlide>
        <div
          className="mfui-beta-tabs__underline"
          slot="wrapper-start"
          style={
            Object {
              "transform": "translateX(0px)",
              "transitionProperty": "none",
              "width": "0px",
            }
          }
        />
        <test-file-stub
          className="mfui-beta-tabs__arrow mfui-beta-tabs__arrow_prev mfui-beta-tabs__arrow_hide"
          onClick={[Function]}
        />
        <test-file-stub
          className="mfui-beta-tabs__arrow mfui-beta-tabs__arrow_next"
          onClick={[Function]}
        />
      </Swiper>
    </div>
  </div>
  <div
    className="mfui-beta-tabs__panel"
    key=".0"
  >
    <Tab
      href="www.test.com"
      icon={<test-file-stub />}
      title="title 1"
    >
      1
    </Tab>
  </div>
  <div
    className="mfui-beta-tabs__panel mfui-beta-tabs__panel_current"
    key=".1"
  >
    <Tab
      renderTabWrapper={[Function]}
      title="title 2"
    >
      2
    </Tab>
  </div>
</div>
`;
